File: tests/robot-cpp/subsystems/elevator.cpp
#include "robot-cpp/subsystems/elevator.hpp"

#include <wpi/system/RobotController.hpp>
#include <wpi/math/controller/PIDController.hpp>
#include <wpi/smartdashboard/SmartDashboard.hpp>
#include <rev/config/SparkMaxConfig.h>

#include "robot-cpp/subsystems/ports.hpp"

namespace {
constexpr double kP = 5.0;
constexpr double kI = 0.0;
constexpr double kD = 0.0;

constexpr double kElevatorGearing = 10.0;
constexpr wpi::units::meter_t kElevatorDrumRadius = 2_in;
constexpr wpi::units::kilogram_t kCarriageMass = 4.0_kg;

constexpr wpi::units::meter_t kMinElevatorHeight = 0_in;
constexpr wpi::units::meter_t kMaxElevatorHeight = 50_in;

wpi::math::DCMotor kElevatorGearbox = wpi::math::DCMotor::Vex775Pro(4);
}  // namespace

Elevator::Elevator()
    : m_motor{0, kElevatorMotorPort, rev::spark::SparkMax::MotorType::kBrushless},
      m_encoder(m_motor.GetEncoder()),
      m_controller(m_motor.GetClosedLoopController()),
      m_elevatorSim(kElevatorGearbox, kElevatorGearing, kCarriageMass,
                    kElevatorDrumRadius, kMinElevatorHeight, kMaxElevatorHeight,
                    true, wpi::units::meter_t{0}) {
  rev::spark::SparkMaxConfig motorConfig;
  motorConfig.closedLoop.P(kP);
  motorConfig.closedLoop.I(kI);
  motorConfig.closedLoop.D(kD);

  m_motor.Configure(motorConfig, rev::ResetMode::kNoResetSafeParameters,
                    rev::PersistMode::kPersistParameters);
}

// void Elevator::Log() {
//   wpi::SmartDashboard::PutNumber("Elevator Height (m)",
//                                  GetElevatorHeight().to<double>());
// }

// wpi::units::meter_t Elevator::GetElevatorHeight() {
//   return wpi::units::meter_t{m_encoder.GetPosition()};
// }

bool Elevator::IsAtHeight() {
  return (GetElevatorHeight() - m_setpoint) < wpi::units::inch_t(2);
}

void Elevator::GoToHeight(wpi::units::meter_t height) {
  m_setpoint = height;
  m_controller.SetSetpoint(height.to<double>(),
                           rev::spark::SparkLowLevel::ControlType::kPosition);
}

void Elevator::Periodic() { Log(); }

void Elevator::SimulationPeriodic() {
  m_elevatorSim.SetInput(Eigen::Vector<double, 1>(
      m_motor.GetThrottle() * wpi::RobotController::GetInputVoltage()));
  m_elevatorSim.Update(20_ms);
  m_encoderPositionSim.Set(m_elevatorSim.GetPosition().to<double>());
}

void Elevator::Stop() { m_motor.SetThrottle(0); }

File: tests/robot-java/src/main/java/frc/robot/subsystems/Elevator.java
package frc.robot.subsystems;

import com.revrobotics.PersistMode;
import com.revrobotics.RelativeEncoder;
import com.revrobotics.ResetMode;
import com.revrobotics.spark.SparkBase;
import com.revrobotics.spark.SparkClosedLoopController;
import com.revrobotics.spark.SparkMax;
import com.revrobotics.spark.config.SparkMaxConfig;
import edu.wpi.first.hal.SimDouble;
import edu.wpi.first.math.system.plant.DCMotor;
import edu.wpi.first.math.util.Units;
import edu.wpi.first.wpilibj.RobotBase;
import edu.wpi.first.wpilibj.RobotController;
import edu.wpi.first.wpilibj.simulation.ElevatorSim;
import edu.wpi.first.wpilibj.simulation.SimDeviceSim;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.Subsystem;

public class Elevator implements Subsystem {
  private static final double kP = 4;
  private static final double kI = 0.0;
  private static final double kD = 0.0;

  private static final double kElevatorGearing = 10.0;
  private static final double kElevatorDrumRadius = Units.inchesToMeters(2);
  private static final double kCarriageMass = 4.0;

  private static final DCMotor kElevatorGearbox = DCMotor.getVex775Pro(4);
  private static final double kMinElevatorHeight = Units.inchesToMeters(0);
  private static final double kMaxElevatorHeight = Units.inchesToMeters(50);

  private static final double kArmEncoderDistPerPulse = 2.0 * 3.14 * kElevatorDrumRadius / 4096.0;

  private final SparkMax m_motor;
  private final RelativeEncoder m_encoder;

  private final SparkClosedLoopController m_pidController;

  // Sim
  private SimDouble m_encoderPositionSim;
  private ElevatorSim m_elevatorSim;

  private double m_goalHeight;

  /** Create a new elevator subsystem. */
  @SuppressWarnings("this-escape")
  public Elevator() {
    m_motor = new SparkMax(PortMap.kElevatorMotorPort, SparkMax.MotorType.kBrushless);
    SparkMaxConfig motorConfig = new SparkMaxConfig();
    motorConfig.encoder.positionConversionFactor(kArmEncoderDistPerPulse);
    motorConfig.closedLoop.p(kP);
    motorConfig.closedLoop.i(kI);
    motorConfig.closedLoop.d(kD);
    m_motor.configure(motorConfig, ResetMode.kResetSafeParameters, PersistMode.kPersistParameters);
    m_encoder = m_motor.getEncoder();
    m_pidController = m_motor.getClosedLoopController();

    if (RobotBase.isSimulation()) {
      SimDeviceSim deviceSim = new SimDeviceSim("SPARK MAX [" + m_motor.getDeviceId() + "]");
      m_encoderPositionSim = deviceSim.getDouble("Position");
      m_elevatorSim =
          new ElevatorSim(
              kElevatorGearbox,
              kElevatorGearing,
              kCarriageMass,
              kElevatorDrumRadius,
              kMinElevatorHeight,
              kMaxElevatorHeight,
              true,
              0);
    }
  }

  public void log() {
    SmartDashboard.putNumber("Elevator Height", m_encoder.getPosition());
  }

  public void goToHeight(double height) {
    m_goalHeight = height;
    m_pidController.setSetpoint(height, SparkBase.ControlType.kPosition);
  }

  public boolean isAtHeight() {
    return Math.abs(m_goalHeight - m_encoder.getPosition()) < 0.05;
  }

  @Override
  public void periodic() {
    log();
  }

  @Override
  public void simulationPeriodic() {
    m_elevatorSim.setInput(m_motor.get() * RobotController.getInputVoltage());
    m_elevatorSim.update(0.02);
    m_encoderPositionSim.set(m_elevatorSim.getPositionMeters());
  }

  public void stop() {
    m_motor.set(0);
  }
}
